File: ctxbisturi/session.py
"""Parse a Claude Code session JSONL, group it into turns, and cut turns safely.

Claude Code stores each session as a JSONL file under
``~/.claude/projects/<slugged-cwd>/<session-id>.jsonl``. Every line is one JSON
object. The ones that matter for context are ``user`` and ``assistant`` rows,
which form a linear chain through ``uuid`` / ``parentUuid``. Other row types
(``attachment``, ``mode``, ``ai-title``, ``file-history-snapshot`` …) hang off
that chain or are standalone metadata.

A *turn* here is one real user prompt plus everything that follows it — its
attachments, the assistant replies, tool-result rows, and trailing metadata —
up to the next real user prompt. That is the unit the TUI lets you cut.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

# Row types that carry a place in the user/assistant conversation chain.
CHAIN_TYPES = ("user", "assistant")


@dataclass
class Entry:
    """One physical line of the JSONL file."""

    index: int          # 0-based line number in the original file
    raw: str            # exact original text (so untouched lines round-trip byte-for-byte)
    obj: dict[str, Any]  # parsed JSON (empty dict if the line failed to parse)

    @property
    def type(self) -> str:
        return self.obj.get("type", "")

    @property
    def uuid(self) -> str | None:
        return self.obj.get("uuid")

    @property
    def parent_uuid(self) -> str | None:
        return self.obj.get("parentUuid")

    def is_prompt(self) -> bool:
        """True for a *real* user prompt (text), not a tool_result carrier."""
        if self.type != "user":
            return False
        if self.obj.get("isMeta"):
            return False
        content = self.obj.get("message", {}).get("content")
        if isinstance(content, str):
            return content.strip() != ""
        if isinstance(content, list):
            has_text = any(
                isinstance(b, dict) and b.get("type") == "text" and b.get("text", "").strip()
                for b in content
            )
            is_tool_result = any(
                isinstance(b, dict) and b.get("type") == "tool_result" for b in content
            )
            return has_text and not is_tool_result
        return False


@dataclass
class Turn:
    """A real prompt and every entry that belongs to it, in file order."""

    number: int                       # 1-based turn number
    entries: list[Entry] = field(default_factory=list)

    @property
    def prompt_entry(self) -> Entry:
        return self.entries[0]

    @property
    def timestamp(self) -> str:
        return self.prompt_entry.obj.get("timestamp", "")

    @property
    def prompt_text(self) -> str:
        return _text_of(self.prompt_entry.obj)

    def title(self, width: int = 60) -> str:
        text = " ".join(self.prompt_text.split())
        if len(text) > width:
            text = text[: width - 1] + "…"
        return text or "(empty prompt)"

    def preview(self) -> list[str]:
        """Human-readable multi-line summary of everything in the turn."""
        lines: list[str] = []
        for e in self.entries:
            t = e.type
            if t == "user" and e.is_prompt():
                lines.append("── you ──")
                lines.extend(_text_of(e.obj).splitlines() or [""])
            elif t == "user":
                lines.append("  [tool result]")
            elif t == "assistant":
                for label in _assistant_labels(e.obj):
                    lines.append(label)
            elif t == "attachment":
                lines.append("  [attachment]")
            # metadata rows are omitted from the preview
        return lines

    def char_size(self) -> int:
        return sum(len(e.raw) for e in self.entries)

    def token_estimate(self) -> int:
        # ~4 chars per token is the usual rough rule for English/code.
        return self.char_size() // 4


def _text_of(obj: dict[str, Any]) -> str:
    content = obj.get("message", {}).get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for b in content:
            if isinstance(b, dict) and b.get("type") == "text":
                parts.append(b.get("text", ""))
        return "\n".join(parts)
    return ""


def _assistant_labels(obj: dict[str, Any]) -> list[str]:
    content = obj.get("message", {}).get("content")
    out: list[str] = []
    if isinstance(content, str):
        return ["── claude ──", *content.splitlines()]
    if isinstance(content, list):
        for b in content:
            if not isinstance(b, dict):
                continue
            bt = b.get("type")
            if bt == "text" and b.get("text", "").strip():
                out.append("── claude ──")
                out.extend(b.get("text", "").splitlines())
            elif bt == "thinking":
                out.append("  [thinking]")
            elif bt == "tool_use":
                out.append(f"  [tool: {b.get('name', '?')}]")
    return out


class Session:
    """A parsed session file, grouped into turns and cuttable by turn number."""

    def __init__(self, path: str, entries: list[Entry]):
        self.path = path
        self.entries = entries
        self.preamble, self.turns = _group(entries)

    @classmethod
    def load(cls, path: str) -> "Session":
        entries: list[Entry] = []
        with open(path, encoding="utf-8") as fh:
            for i, line in enumerate(fh):
                stripped = line.rstrip("\n")
                if not stripped.strip():
                    continue
                try:
                    obj = json.loads(stripped)
                except json.JSONDecodeError:
                    obj = {}
                entries.append(Entry(index=i, raw=stripped, obj=obj))
        return cls(path, entries)

    def total_tokens(self) -> int:
        return sum(t.token_estimate() for t in self.turns)

    def cut(self, turn_numbers: set[int]) -> list[str]:
        """Return new JSONL lines with the given turns removed and the chain repaired.

        Does not touch disk — callers decide where to write. See :func:`cut_entries`.
        """
        keep_entries = list(self.preamble)
        for turn in self.turns:
            if turn.number not in turn_numbers:
                keep_entries.extend(turn.entries)
        return _rethread(keep_entries)


def _group(entries: list[Entry]) -> tuple[list[Entry], list[Turn]]:
    """Split entries into a leading preamble and one Turn per real prompt."""
    first_prompt = next((i for i, e in enumerate(entries) if e.is_prompt()), None)
    if first_prompt is None:
        return list(entries), []
    preamble = entries[:first_prompt]
    turns: list[Turn] = []
    current: Turn | None = None
    for e in entries[first_prompt:]:
        if e.is_prompt():
            current = Turn(number=len(turns) + 1, entries=[e])
            turns.append(current)
        else:
            assert current is not None
            current.entries.append(e)
    return preamble, turns


def _rethread(entries: list[Entry]) -> list[str]:
    """Repair the uuid chain of a surviving set of entries and serialize to lines.

    Sessions are linear, so after removing a middle block we relink each
    surviving chain message's ``parentUuid`` to the previous surviving chain
    message. Rows (attachments, metadata) that reference a uuid no longer
    present are dropped so a resume never dereferences a hole.
    """
    surviving_uuids = {e.uuid for e in entries if e.type in CHAIN_TYPES and e.uuid}

    out_lines: list[str] = []
    prev_chain_uuid: str | None = None
    for e in entries:
        obj = e.obj
        if not obj:
            out_lines.append(e.raw)
            continue

        if e.type in CHAIN_TYPES:
            # Relink to the previous survivor; untouched if it already matches.
            if obj.get("parentUuid") != prev_chain_uuid:
                obj = dict(obj)
                obj["parentUuid"] = prev_chain_uuid
                out_lines.append(json.dumps(obj, ensure_ascii=False))
            else:
                out_lines.append(e.raw)
            prev_chain_uuid = e.uuid
            continue

        # Non-chain rows: drop if they point at a removed message.
        ref = obj.get("parentUuid")
        leaf = obj.get("leafUuid")
        if ref is not None and ref not in surviving_uuids:
            continue
        if leaf is not None and leaf not in surviving_uuids:
            continue
        out_lines.append(e.raw)

    return out_lines

File: internal/session/session.go
// Package session parses a Claude Code session transcript, groups it into
// turns, and cuts turns while keeping the result resumable.
//
// A Claude Code session is a JSONL file at
// ~/.claude/projects/<slugged-cwd>/<session-id>.jsonl. Each line is one JSON
// object. Objects that carry a "uuid" (user, assistant, attachment, system, …)
// are threaded by "parentUuid". Parallel tool calls, retries, sidechains and
// compaction make the real structure a *forest* — many roots and leaf tips —
// so there is no single reliable "active path" to reason about across all
// sessions.
//
// Instead we treat the file as what it is: an append-ordered (chronological)
// log. Cutting removes the entries of the chosen turns and *relinks* the
// survivors into one clean linear chain — each surviving uuid node points at the
// previous surviving uuid node. That always yields a single-rooted, fully
// connected transcript with no dangling references, whatever the original shape,
// and it resumes correctly whether Claude Code walks parentUuid from the leaf or
// reads the log in order.
//
// A *turn* is one real user prompt plus the entries that follow it up to the
// next real prompt. Injected prompts (hook feedback, skill preambles; isMeta)
// fold into their surrounding turn, so you select whole topics.
package session

import (
	"bufio"
	"encoding/json"
	"os"
	"strings"
)

// Entry is one physical line of the JSONL file. Raw is kept verbatim so
// untouched lines round-trip byte-for-byte; Obj is the parsed form.
type Entry struct {
	Index int
	Raw   string
	Obj   map[string]any
}

func (e Entry) Type() string      { return str(e.Obj["type"]) }
func (e Entry) UUID() string      { return str(e.Obj["uuid"]) }
func (e Entry) Parent() string    { return str(e.Obj["parentUuid"]) }
func (e Entry) HasUUID() bool     { return e.UUID() != "" }
func (e Entry) Timestamp() string { return str(e.Obj["timestamp"]) }

// IsPrompt reports whether this is a real user prompt (text the user typed),
// not a tool_result carrier or an injected/meta message.
func (e Entry) IsPrompt() bool {
	if e.Type() != "user" {
		return false
	}
	if b, _ := e.Obj["isMeta"].(bool); b {
		return false
	}
	switch c := messageContent(e.Obj).(type) {
	case string:
		return strings.TrimSpace(c) != ""
	case []any:
		hasText, isToolResult := false, false
		for _, blk := range c {
			m, ok := blk.(map[string]any)
			if !ok {
				continue
			}
			switch str(m["type"]) {
			case "text":
				if strings.TrimSpace(str(m["text"])) != "" {
					hasText = true
				}
			case "tool_result":
				isToolResult = true
			}
		}
		return hasText && !isToolResult
	}
	return false
}

// Turn is a real prompt and every entry that belongs to it, in file order.
type Turn struct {
	Number  int
	Entries []Entry
}

func (t Turn) Prompt() Entry      { return t.Entries[0] }
func (t Turn) Timestamp() string  { return t.Prompt().Timestamp() }
func (t Turn) PromptText() string { return cleanText(textOf(t.Prompt().Obj)) }

// CharSize is the total raw byte size of the turn (with newlines).
func (t Turn) CharSize() int {
	n := 0
	for _, e := range t.Entries {
		n += len(e.Raw) + 1
	}
	return n
}

// TokenEstimate is a rough ~4-chars-per-token estimate, for relative sizing.
func (t Turn) TokenEstimate() int { return t.CharSize() / 4 }

// Session is a parsed session grouped into a preamble plus turns.
type Session struct {
	Path     string
	Entries  []Entry
	Preamble []Entry
	Turns    []Turn
}

// Load reads and parses a session file.
func Load(path string) (*Session, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var entries []Entry
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 0, 1024*1024), 64*1024*1024)
	i := 0
	for sc.Scan() {
		line := sc.Text()
		if strings.TrimSpace(line) == "" {
			i++
			continue
		}
		var obj map[string]any
		_ = json.Unmarshal([]byte(line), &obj)
		entries = append(entries, Entry{Index: i, Raw: line, Obj: obj})
		i++
	}
	if err := sc.Err(); err != nil {
		return nil, err
	}
	s := &Session{Path: path, Entries: entries}
	s.group()
	return s, nil
}

// group partitions entries into a leading preamble and one Turn per real prompt.
func (s *Session) group() {
	first := -1
	for i, e := range s.Entries {
		if e.IsPrompt() {
			first = i
			break
		}
	}
	if first == -1 {
		s.Preamble = s.Entries
		return
	}
	s.Preamble = s.Entries[:first]
	var cur *Turn
	for _, e := range s.Entries[first:] {
		if e.IsPrompt() {
			s.Turns = append(s.Turns, Turn{Number: len(s.Turns) + 1, Entries: []Entry{e}})
			cur = &s.Turns[len(s.Turns)-1]
		} else if cur != nil {
			cur.Entries = append(cur.Entries, e)
		}
	}
}

// TotalTokens sums the estimated tokens across turns.
func (s *Session) TotalTokens() int {
	n := 0
	for _, t := range s.Turns {
		n += t.TokenEstimate()
	}
	return n
}

// RemovedRun is one contiguous block taken out by a cut, with the file anchor
// needed to splice it back in later.
type RemovedRun struct {
	AnchorBefore string   `json:"anchor_before"` // uuid to re-insert after ("" = front)
	Lines        []string `json:"lines"`         // removed raw lines, in original order
}

// CutResult is the outcome of a cut: the relinked file lines plus what was
// removed (grouped into contiguous runs for restore).
type CutResult struct {
	Lines []string
	Runs  []RemovedRun
}

// Cut removes the given turn numbers and relinks the survivors into a clean
// linear chain. It does not touch disk.
func (s *Session) Cut(cut map[int]bool) CutResult {
	removedIdx := map[int]bool{}
	removedUUID := map[string]bool{}
	for _, t := range s.Turns {
		if !cut[t.Number] {
			continue
		}
		for _, e := range t.Entries {
			removedIdx[e.Index] = true
			if e.HasUUID() {
				removedUUID[e.UUID()] = true
			}
		}
	}

	survivors := make([]Entry, 0, len(s.Entries))
	for _, e := range s.Entries {
		if !removedIdx[e.Index] {
			survivors = append(survivors, e)
		}
	}
	lines := relink(survivors, removedUUID)

	return CutResult{Lines: lines, Runs: buildRuns(s.Entries, removedIdx)}
}

// relink walks entries in order and rethreads every uuid-bearing node to the
// previous surviving uuid node, so the chain is linear and gap-free. Untouched
// links are left byte-identical. Metadata rows that reference a removed uuid are
// dropped so nothing dangles.
func relink(entries []Entry, removedUUID map[string]bool) []string {
	out := make([]string, 0, len(entries))
	prev := ""
	for _, e := range entries {
		if e.Obj == nil {
			out = append(out, e.Raw)
			continue
		}
		if e.HasUUID() {
			desired := prev
			if e.Parent() != desired {
				clone := cloneObj(e.Obj)
				if desired == "" {
					clone["parentUuid"] = nil
				} else {
					clone["parentUuid"] = desired
				}
				out = append(out, marshal(clone))
			} else {
				out = append(out, e.Raw)
			}
			prev = e.UUID()
			continue
		}
		// No-uuid metadata: drop if it points at a removed message.
		if leaf := str(e.Obj["leafUuid"]); leaf != "" && removedUUID[leaf] {
			continue
		}
		if p := str(e.Obj["parentUuid"]); p != "" && removedUUID[p] {
			continue
		}
		out = append(out, e.Raw)
	}
	return out
}

// buildRuns groups removed entries into contiguous file-order runs and records,
// for each, the uuid of the surviving node immediately before it (the anchor to
// re-insert after on restore).
func buildRuns(all []Entry, removedIdx map[int]bool) []RemovedRun {
	var runs []RemovedRun
	var cur *RemovedRun
	lastSurvivorUUID := ""
	for _, e := range all {
		if removedIdx[e.Index] {
			if cur == nil {
				runs = append(runs, RemovedRun{AnchorBefore: lastSurvivorUUID})
				cur = &runs[len(runs)-1]
			}
			cur.Lines = append(cur.Lines, e.Raw)
			continue
		}
		cur = nil
		if e.HasUUID() {
			lastSurvivorUUID = e.UUID()
		}
	}
	return runs
}

// Relink rethreads a set of raw JSONL lines into a clean linear chain. Used on
// restore after removed blocks are spliced back in.
func Relink(rawLines []string) []string {
	entries := make([]Entry, 0, len(rawLines))
	for i, line := range rawLines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var obj map[string]any
		_ = json.Unmarshal([]byte(line), &obj)
		entries = append(entries, Entry{Index: i, Raw: line, Obj: obj})
	}
	return relink(entries, map[string]bool{})
}

// --- helpers ---

func messageContent(obj map[string]any) any {
	msg, ok := obj["message"].(map[string]any)
	if !ok {
		return nil
	}
	return msg["content"]
}

func textOf(obj map[string]any) string {
	switch c := messageContent(obj).(type) {
	case string:
		return c
	case []any:
		var parts []string
		for _, blk := range c {
			if m, ok := blk.(map[string]any); ok && str(m["type"]) == "text" {
				parts = append(parts, str(m["text"]))
			}
		}
		return strings.Join(parts, "\n")
	}
	return ""
}

func str(v any) string { s, _ := v.(string); return s }

func cloneObj(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

func marshal(m map[string]any) string {
	b, _ := json.Marshal(m)
	return string(b)
}
